### ISSUE TITLE
feat: persist agent run traces for fine-tuning datasets

### ISSUE THREAD

**Issue by ponderingdemocritus** [BibliothecaDAO/eternum#4446]
## Overview

We need to capture complete agent run traces from Eternum so every run can be turned into training data for later fine-tuning, evals, ranking datasets, and post-hoc analysis.

This should **not** start as a PI core change.

The valuable data is not just generic PI agent output. It is the full Eternum trajectory:
- world state before a turn
- rendered prompt and active task context
- assistant output
- tool calls and tool results
- chosen game action and tx result
- world state after the action
- operator interventions
- eventual outcomes and derived reward/quality labels

That data is game-specific, so the first implementation should live in the Eternum stack. If the generic tracing seam is useful across projects, we can extract it upstream into PI later.

## Problem

Today we have lightweight observability, but not a training-grade trace pipeline:
- `packages/game-agent` has a markdown decision recorder, but it is summary-oriented and lossy.
- `client/apps/onchain-agent` subscribes to PI agent events and logs them to stdout, but those events are not persisted as structured run data.
- We do not persist per-turn state snapshots, prompt inputs, tool trajectories, or post-action outcomes in a machine-consumable format.
- We cannot reliably reconstruct a run into SFT, preference, or tool-use datasets.

That leaves us unable to:
- build high-quality supervised fine-tuning corpora
- compare good vs bad trajectories
- train on real action selection sequences
- reproduce and audit model behavior after a run
- evaluate prompt/model changes against historical traces

## Decision Summary

Implement this in two layers:

### Layer 1: Canonical tracing contract in our agent abstraction
Add a reusable trace sink interface at the `packages/game-agent` boundary so game-agent users can emit structured lifecycle events without baking Eternum-specific storage into the package.

### Layer 2: Eternum-specific trace capture and storage in `client/apps/onchain-agent`
Implement the first real sink in `client/apps/onchain-agent`.
This sink will persist newline-delimited JSON event streams, content-addressed state blobs, and run metadata suitable for offline dataset generation.

### Deferred extraction
If this seam proves broadly useful, extract the generic event hook shape into PI later. Do **not** block Eternum on an upstream PI package change.

## Goals

1. Persist every agent run as a structured, replayable trace.
2. Capture enough information to derive:
   - SFT datasets
   - tool-use datasets
   - preference/ranking datasets
   - eval fixtures
3. Keep raw capture append-only and lossless enough that future dataset transforms do not need another product migration.
4. Avoid coupling raw trace capture to any single training format.
5. Support local dev runs first, then durable remote upload.

## Non-Goals

- No direct fine-tuning job orchestration in this issue.
- No model-training pipeline implementation in this issue.
- No PI upstream package changes in this issue.
- No browser-side tracing for player chat UI in this issue.
- No registry/analytics dashboard beyond basic queryability.

## Primary Users

- ML/agent engineers building SFT and ranking datasets.
- gameplay engineers debugging action quality and tool usage.
- operators comparing model versions and prompt versions across runs.

## Success Metrics

- 100% of agent turns emit a structured trace event sequence from turn start to turn end.
- 100% of executed game actions have linked tool call, action payload, and result records.
- A single completed run can be transformed into an offline dataset without reading stdout logs.
- We can reconstruct at least one end-to-end run into a training example set covering state -> reasoning -> tool call -> action -> outcome.

### `packages/game-agent`
- `createGameAgent()` already owns enqueue semantics, tick integration, runtime config updates, and the decision recorder seam.
- This is the correct place for a generic trace sink contract.

### `client/apps/onchain-agent`
- already subscribes to PI agent events
- already owns Eternum world bootstrap, tool wiring, and action execution
- already knows the world id, chain, account, tick cadence, and data dir

### Current gaps
- decision recorder is too lossy
- stdout logging is not a dataset
- world state is not persisted as structured, versioned blobs

### 1. Add a generic trace sink contract in `packages/game-agent`

Add a new tracing module with interfaces similar to:

```ts
export interface AgentTraceSink {
  onRunStart(event: RunStartEvent): Promise<void>;
  onTurnStart(event: TurnStartEvent): Promise<void>;
  onPromptBuilt(event: PromptBuiltEvent): Promise<void>;
  onModelEvent(event: ModelEvent): Promise<void>;
  onToolCall(event: ToolCallEvent): Promise<void>;
  onToolResult(event: ToolResultEvent): Promise<void>;
  onActionSubmitted(event: ActionSubmittedEvent): Promise<void>;
  onStateObserved(event: StateObservedEvent): Promise<void>;
  onTurnEnd(event: TurnEndEvent): Promise<void>;
  onRunEnd(event: RunEndEvent): Promise<void>;
}
```

`createGameAgent()` should accept an optional `traceSink` and emit lifecycle events at the orchestration layer.

Important rule:
- `packages/game-agent` defines the event contract.
- `packages/game-agent` does **not** own Eternum storage, upload policy, or dataset export logic.

### 2. Implement `FileTraceSink` in `client/apps/onchain-agent`

Create an Eternum-specific sink that writes:
- `run.json` manifest
- `events.jsonl` append-only event stream
- `blobs/<sha256>.json.gz` content-addressed state/prompt/tool-result blobs
- optional derived summaries for operator/debug use

Suggested run directory:

```text
~/.axis/runs/<run-id>/
  run.json
  events.jsonl
  blobs/
  summary.json
```

### 3. Use content-addressed blobs for large payloads

Do not inline giant repeated payloads in every event.

Persist these as blobs keyed by digest:
- pre-turn normalized world state
- post-turn normalized world state
- rendered prompt
- long assistant responses
- large tool results

Each event references the blob via hash and metadata:
- `blob_id`
- `content_type`
- `compression`
- `byte_length`

### 4. Normalize the world state for training use

Capture two representations:

#### Raw observation blob
The exact state object returned by the adapter or map pipeline.

#### Training projection blob
A normalized, stable schema intended for downstream dataset generation.

This prevents training data from depending directly on transient internal runtime shapes.

### 5. Keep raw capture and dataset export separate

Raw traces are source-of-truth.

Add a separate export step later to generate:
- SFT JSONL
- preference pairs
- eval fixtures
- action outcome summaries

This issue should define the raw data shape so future exports are straightforward.

### `packages/game-agent`
Add:
- `trace.ts` or equivalent tracing module
- exported event types
- exported `AgentTraceSink` interface
- optional `traceSink?: AgentTraceSink` in `CreateGameAgentOptions`

### `client/apps/onchain-agent`
Add:
- `tracing/file-trace-sink.ts`
- `tracing/blob-store.ts`
- `tracing/event-types.ts` if Eternum-specific extensions are needed
- `tracing/run-context.ts`
- `tracing/export/` reserved for later dataset transforms

### Config additions
Add env/config support for:
- `TRACE_ENABLED=true|false`
- `TRACE_DIR=<path>`
- `TRACE_CAPTURE_PROMPTS=true|false`
- `TRACE_CAPTURE_FULL_STATE=true|false`
- `TRACE_GZIP_BLOBS=true|false`
- `TRACE_UPLOAD_ENABLED=true|false` (stubbed for later)
- `TRACE_REDACT_SECRETS=true|false`

Defaults:
- tracing enabled in local/dev and server modes
- prompts captured by default
- full state captured by default
- uploads disabled by default
- secret redaction enabled by default

### Run-level metadata
Persist once in `run.json`:

```json
{
  "run_id": "uuid",
  "agent_id": "string",
  "owner_player_id": "string|null",
  "world_id": "string",
  "world_name": "string|null",
  "chain": "slot|slottest|sepolia|mainnet|local",
  "model_provider": "anthropic|openai|groq|x402|...",
  "model_id": "string",
  "tick_interval_ms": 60000,
  "prompt_version": "git sha or explicit version",
  "git_sha": "string",
  "started_at": "iso8601",
  "ended_at": "iso8601|null",
  "end_reason": "completed|aborted|crashed|timeout|null"
}
```

### Event stream schema
Each line in `events.jsonl`:

```json
{
  "event_id": "uuid",
  "run_id": "uuid",
  "turn_id": "uuid|null",
  "seq": 42,
  "ts": "iso8601",
  "type": "turn_start",
  "payload": {}
}
```

### Required event types
- `run_start`
- `state_observed`
- `turn_start`
- `prompt_built`
- `agent_start`
- `assistant_message`
- `tool_call_start`
- `tool_call_end`
- `action_submitted`
- `action_result`
- `turn_end`
- `run_end`
- `error`
- `operator_message` (for human interventions)
- `config_change`

### Required linkages
Every event must carry enough IDs to correlate:
- run
- turn
- tool invocation
- action execution
- snapshot before/after

## What Must Be Captured Per Turn

For every turn we need:
- turn number / game tick
- pre-turn normalized state reference
- rendered prompt reference
- active memory/task file references or hashes
- assistant output
- tool call sequence
- tool arguments
- tool results
- action payloads
- tx hash / execution result
- post-turn normalized state reference
- latency and error metadata

## Redaction / Privacy Rules

Never persist secrets in traces.

Must redact or exclude:
- API keys
- session tokens
- auth URLs containing credentials
- private keys
- cookies or browser auth state

Default policy:
- keep prompts, tool calls, world state, and outputs
- scrub env-derived secrets before persistence

### In `packages/game-agent`
Emit generic lifecycle events around:
- run start
- prompt enqueue
- tick prompt generation
- turn start/end
- tool execution start/end
- agent output receipt
- runtime config changes

### In `client/apps/onchain-agent`
Emit Eternum-specific events around:
- world bootstrap
- map snapshot observation
- action submission
- tx result
- automation interactions if they materially affect training context
- operator stdin prompts / steering

### At the adapter/tool boundary
Capture actual game semantics here, not just model chatter:
- chosen action type
- params
- simulation result
- execution result
- tx hash
- failure mode

### Phase 1 storage
Local append-only spool on disk:
- easiest to inspect
- resilient in dev
- good enough to validate schema

### Phase 2 storage
Asynchronous upload to durable object storage:
- blob payloads to S3/R2/GCS
- metadata index to Postgres or ClickHouse

This issue only needs to define the seam so upload can be added without changing event shape.

## Dataset Derivation Plan

The raw trace should support later derivation into:

### SFT examples
Input:
- normalized state
- active goals/memory
- prompt context

Target:
- assistant reasoning or action proposal

### Tool-use examples
Input:
- prompt plus prior context

Target:
- tool call sequence and arguments

### Preference / ranking examples
Input:
- two candidate trajectories or decisions

Label:
- which outcome was better

### Eval fixtures
Frozen replay cases with expected action quality or tool usage patterns.

### Phase 1
Add local file-based trace capture behind a config flag.

### Phase 2
Enable tracing by default for `onchain-agent` dev/server runs.

### Phase 3
Build offline inspection scripts to validate completeness and dataset usefulness.

### Phase 4
Add optional remote upload.

### Unit tests: `packages/game-agent`
1. emits `run_start` and `run_end` when the agent lifecycle starts/stops
2. emits `turn_start` and `turn_end` for each prompt-driven turn
3. emits tool call start/end events with stable invocation IDs
4. emits error events when prompt or tool execution fails
5. does not break existing agent behavior when no trace sink is provided
6. preserves enqueue/follow-up semantics with tracing enabled

### Unit tests: `client/apps/onchain-agent` tracing
1. `FileTraceSink` writes valid NDJSON events in order
2. large payloads are stored as blobs and referenced by hash
3. duplicate blob content deduplicates to one blob id
4. prompt capture respects redaction rules
5. run manifest is finalized correctly on clean shutdown
6. crash/interruption writes `run_end` or recovery metadata correctly when possible

### Integration tests
1. one complete agent run produces:
   - `run.json`
   - `events.jsonl`
   - at least one state blob
   - at least one prompt blob
2. a run containing tool calls links tool events to action result events
3. a run with human steering captures `operator_message` plus subsequent model/tool behavior
4. a failed tool execution still results in a valid, queryable trace

### Dataset validation tests
1. trace export can derive at least one SFT example from a completed run
2. trace export can derive at least one tool-use example from a completed run
3. exporter fails loudly when a trace is structurally incomplete

## Acceptance Criteria

- A completed agent run is persisted as structured trace data without relying on stdout logs.
- Every turn has linked prompt, state, tool, and result records.
- Large repeated payloads are blob-referenced rather than duplicated inline.
- Secrets are not present in persisted traces.
- Tracing can be enabled/disabled via config.
- Existing agent behavior remains unchanged when tracing is disabled.
- At least one sample trace can be transformed into a training-ready dataset offline.

### Risk: trace volume explodes
Mitigation:
- blob deduplication
- gzip compression
- configurable capture levels

### Risk: trace schema becomes too PI-specific or too Eternum-specific
Mitigation:
- keep generic lifecycle contract in `packages/game-agent`
- keep storage and game semantics in `client/apps/onchain-agent`

### Risk: prompts/state contain sensitive data
Mitigation:
- explicit redaction layer
- safe defaults
- test fixtures that assert secrets are scrubbed

### Risk: capture overhead affects runtime
Mitigation:
- append-only writes
- optional async flush/buffering
- configurable capture levels

### Risk: traces are not useful for training later
Mitigation:
- preserve raw truth now
- defer lossy training transforms to separate export steps
- validate by deriving sample SFT/tool-use outputs before calling the work done

## Explicit Defaults Chosen

- Build this in Eternum first, not PI core.
- Put the generic trace sink interface in `packages/game-agent`.
- Put the first concrete sink implementation in `client/apps/onchain-agent`.
- Store raw traces as NDJSON plus content-addressed blobs.
- Separate raw trace capture from training dataset export.
- Enable local tracing first; remote upload comes later.

## Suggested Implementation Order

1. Define generic trace interfaces and event contracts in `packages/game-agent`.
2. Add no-op-compatible instrumentation to `createGameAgent()`.
3. Implement local `FileTraceSink` in `client/apps/onchain-agent`.
4. Instrument Eternum-specific state/action boundaries.
5. Add fixture-based integration tests.
6. Add an offline dataset derivation smoke test.
7. Add optional remote upload in a follow-up issue.
